diff --git a/SConstruct b/SConstruct
index e8107ea..a678711 100755
--- a/SConstruct
+++ b/SConstruct
@@ -265,6 +265,10 @@ main.Append(CPPPATH=[Dir('ext')])
 # Add shared top-level headers
 main.Prepend(CPPPATH=Dir('include'))
 
+if 'CONDA_PREFIX' in environ:
+    main.Prepend(CPPPATH=[join(environ['CONDA_PREFIX'], 'include')])
+    main.Prepend(LIBPATH=[join(environ['CONDA_PREFIX'], 'lib')])
+
 
 ########################################################################
 #
@@ -285,7 +289,8 @@ main['TCMALLOC_CCFLAGS'] = []
 
 CXX_version = readCommand([main['CXX'], '--version'], exception=False)
 
-main['GCC'] = CXX_version and CXX_version.find('g++') >= 0
+main['GCC'] = CXX_version and (CXX_version.find('g++') >= 0 or
+                               CXX_version.find('gcc') >= 0)
 main['CLANG'] = CXX_version and CXX_version.find('clang') >= 0
 if main['GCC'] + main['CLANG'] > 1:
     error('Two compilers enabled at once?')
